Created a more fine grained control over chat and log channel using the config file

v-tron opened the pull request:
I added two categories to the config file to give the user more fine grained control over what the bot sends to each channel, potentially significantly reducing clutter in the log channel (like user messages, which clutter the log when you just want quick info over who joined when for example) and also freeing up the chat channel from unnecessary server messages (like starting/stopping server, automatic periodic rcon messages i.e. for backup routines and more).

I needed to put a lot of if statements in the Handler which I did not really like, but the alternative would be rewriting a lot of stuff and that was even less appealing. The result is effectively deprecating the sendToAllChannels function. This can still be a useful function, I just did not see how without creating more bloated if statements.

The jar built without issues and everything seems to be working correctly on my own server.

v-tron committed:
Changed config file and java files to support fine-grained control over chat and log channel.

v-tron committed:
removed the api import, whereever that came from

v-tron committed:
Fixed minor typo and changed default config slightly.

v-tron committed:
commented out sendToAllChannels function because it is not currently used

v-tron committed:
Fixed error in Code that enables Server starting message with Server Stopping setting

thecatcore reviewed (approved):
Thanks for fixing it, I will merge it when I come back home in a few days.

v-tron committed:
Added discord bot commands "!status", "!uptime", "!playercount" and "!playerlist".
Implemented so that commands do not get sent to minecraft server chat as to not annoy players playing.

Removed customChannelDescription feature due to discords API rate-limits.
Could be changed to display every 5 minutes but that is pretty useless.

Added options deathMsgPrefix and deathMessagePostfix to message config.

v-tron committed:
Removed Arraylist I forgot to remove after implementing differently

v-tron committed:
Created Features und Installation documentation

v-tron committed:
minor fix

v-tron committed:
more small fixes

v-tron committed:
ok this is the last one I promise lol

AlexanderRitter02 commented:
@v-tron In post- and prefixes you can't use the `%player`. You may want to add this to this pull request as well.
Would be needed for https://github.com/arthurbambou/Fabric---Discord-Link/issues/81#issuecomment-667111563

AlexanderRitter02 commented on src/main/java/fr/arthurbambou/fdlink/discordstuff/DiscordBot.java:
```diff
@@ -86,40 +85,44 @@ public DiscordBot(String token, FDLink.Config config) {
+            if (this.config.minecraftToDiscord.chatChannels.serverStopMessage || this.config.minecraftToDiscord.logChannels.serverStopMessage) {
+                List<CompletableFuture<Message>> requests = new ArrayList<>();
+                if(this.config.minecraftToDiscord.chatChannels.serverStopMessage) requests.add(sendToChatChannels(config.minecraftToDiscord.messages.serverStopped));
+                if(this.config.minecraftToDiscord.logChannels.serverStopMessage) requests.add(sendToLogChannels(config.minecraftToDiscord.messages.serverStopped));
```
This added line **(line 111)** produces a `NullPointerException` when shutting down the server.
When I comment it out, the exception doesn't happen anymore.

```
ERROR MinecraftServer Exception stopping the server
 java.lang.NullPointerException
	at fr.arthurbambou.fdlink.discordstuff.DiscordBot.lambda$new$5(DiscordBot.java:113)
	at net.minecraft.server.MinecraftServer.handler$zzp000$afterShutdownServer(MinecraftServer.java:1838)
	at net.minecraft.server.MinecraftServer.shutdown(MinecraftServer.java:618)
	at net.minecraft.server.dedicated.MinecraftDedicatedServer.shutdown(MinecraftDedicatedServer.java:500)
	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:699)
	at net.minecraft.server.MinecraftServer.method_29739(MinecraftServer.java:226)
	at java.lang.Thread.run(Thread.java:748)
```

v-tron replied:
Yeah, I see why that would happen. Thanks for the heads up.

v-tron replied:
Is this from the minecraft logs? I cannot seem to reproduce the Exception on my end.

AlexanderRitter02 replied:
This is straight from the _server console_ (also in latest.log). Happens every time for me when running `/stop`, even if run just after startup.
I haven't changed the default config, only added `token` and `chatChannels`. I have no logChannels.
Only other mod on the server is `fabric-api-0.16.2+build.385-1.16.1.jar`.

Currently there is nothing more I can say, however if I comment the line out and recompile, the error doesn't occur anymore.

thecatcore replied:
@v-tron do you set the log channel’s booleans to false if there is no log channel in the config?

v-tron replied:
@arthurbambou oh yea, I forgot to do an actual check for hasLogChannels because I thought that the sendToLogChannels function does that. But it actually returns null on no log Channels so that causes the exception I guess

AlexanderRitter02 commented on src/main/java/fr/arthurbambou/fdlink/discordstuff/DiscordBot.java:
```diff
@@ -166,40 +204,51 @@ public DiscordBot(String token, FDLink.Config config) {
-                            "player count : %d/%d,\n" +
-                                    "uptime : %d h %d min %d second",
-                            playerNumber, maxPlayer, uptimeH, uptimeM, uptimeS
+                if(this.config.minecraftToDiscord.chatChannels.customChannelDescription){
```
For now it's good to have that removed since you could be API banned by Discord.
However we should look at reimplementing it with a proper delay later (#92 discussion)

v-tron replied:
Doing 5 or 10 Minutes is an easy fix. I will readd the feature since it seems people still want it, even though it's not accurate information.

v-tron committed:
Readded customChannelDescription with a 5 minute cycle to avoid Discord API rate-limit

Commit:
finxed nullPointerException on shutdown
